subsys/net/lib/aws_iot/src/aws_iot.c

```diff
@@ -0,0 +1,679 @@
+{
+	if (topic_list_count == 0) {
+		return -EMSGSIZE;
+	}
+
+	app_topic_data.list =
+		(struct mqtt_topic *)malloc(sizeof(struct mqtt_topic) * topic_list_count);
+
+
+	for (int i = 0; i < topic_list_count; i++) {
+		app_topic_data.list[i].topic.utf8 = topic_list[i].str;
+		app_topic_data.list[i].topic.size = topic_list[i].len;
+		app_topic_data.list[i].qos = MQTT_QOS_1_AT_LEAST_ONCE;
+	}
+
+	app_topic_data.list_count = topic_list_count;
+
+	return 0;
+}
+
+static int data_publish(struct mqtt_client *const c, enum mqtt_qos qos,
+			char *data, size_t len, char *topic)
+{
+	struct mqtt_publish_param param;
+
+	param.message.topic.qos = qos;
+	param.message.topic.topic.utf8 = topic;
+	param.message.topic.topic.size = strlen(topic);
+	param.message.payload.data = data;
+	param.message.payload.len = len;
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0;
+	param.retain_flag = 0;
+
+	LOG_DBG("Publishing to topic: %s",
+		log_strdup(param.message.topic.topic.utf8));
+
+	return mqtt_publish(c, &param);
+}
+
+static int topic_subscribe(void)
+{
+	int err;
+
+	if (app_topic_data.list_count > 0) {
+		const struct mqtt_subscription_list app_sub_list = {
+			.list = app_topic_data.list,
+			.list_count = app_topic_data.list_count,
+			.message_id = sys_rand32_get()
+		};
+
+		for (size_t i = 0; i < app_sub_list.list_count; i++) {
+			LOG_DBG("Subscribing to application spesific topic: %s",
+			log_strdup(app_sub_list.list[i].topic.utf8));
+		}
+
+		err = mqtt_subscribe(&client, &app_sub_list);
+		if (err) {
+			LOG_ERR("Could not subscribe to application spesific topics, error: %d", err);
+		}
+	}
+
+	if (ARRAY_SIZE(aws_iot_rx_list) > 0) {
+		const struct mqtt_subscription_list aws_sub_list = {
+			.list = (struct mqtt_topic *)&aws_iot_rx_list,
+			.list_count = ARRAY_SIZE(aws_iot_rx_list),
+			.message_id = sys_rand32_get()
+		};
+
+		for (size_t i = 0; i < aws_sub_list.list_count; i++) {
+			LOG_DBG("Subscribing to aws spesific topic: %s",
+			log_strdup(aws_sub_list.list[i].topic.utf8));
+		}
+
+		err = mqtt_subscribe(&client, &aws_sub_list);
+		if (err) {
+			LOG_ERR("Could not subscribe to AWS spesific topics, error: %d", err);
+		}
+	}
+
+	return err;
+}
+
+static int publish_get_payload(struct mqtt_client *const c, size_t length)
+{
+	if (length > sizeof(payload_buf)) {
+		return -EMSGSIZE;
+	}
+
+	return mqtt_readall_publish_payload(c, payload_buf, length);
+}
+
+static void mqtt_evt_handler(struct mqtt_client *const c,
+			     const struct mqtt_evt *mqtt_evt)
+{
+	int err;
+	struct cloud_backend_config *config = aws_iot_backend->config;
+	struct cloud_event cloud_evt = { 0 };
+
+#if defined(CONFIG_AWS_FOTA)
+	err = aws_fota_mqtt_evt_handler(c, mqtt_evt);
+	if (err > 0) {
+		/* Event handled by FOTA library so we can skip it */
+		return;
+	} else if (err < 0) {
+		LOG_ERR("aws_fota_mqtt_evt_handler, error: %d", err);
+		cloud_evt.type = CLOUD_EVT_ERROR;
+		cloud_notify_event(aws_iot_backend, &cloud_evt,
+				   config->user_data);
```

Comment:
Shouldn't we quit on error as well? I guess that if `aws_fota` threw an error, it has handled the event.

Reply:
This is now changed according to the ```aws_fota``` example where the device disconnects from MQTT upon a ```FOTA_ERROR``` event.